File: java/3134. Find the Median of the Uniqueness Array.java
class Solution {
    public int medianOfUniquenessArray(int[] nums) {
        long total = ((long)nums.length * (nums.length+1))/2;
        int left = 1;
        int right = nums.length;
        int answer = 1;
        while (left <= right) {
            int mid = (left + right) / 2;
            long count = check(nums, mid);
            if (count * 2 >= total) {
                answer = mid;
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        return answer;
    }
    private long check(int[] nums, int mid) {
        long count = 0;
        Map<Integer, Integer> freq = new HashMap<>();
        int size = 0;
        int left = 0;
        for (int right = 0; right < nums.length; right++) {
            if (!freq.containsKey(nums[right]) || freq.get(nums[right]) == 0) size++;
            freq.put(nums[right], freq.getOrDefault(nums[right], 0) + 1);
            while (size > mid) {
                freq.put(nums[left], freq.get(nums[left]) - 1);
                if (freq.get(nums[left]) == 0) size--;
                left++;
            }
            count += right - left + 1;
        }
        return count;
    }
}

File: go/3134. Find the Median of the Uniqueness Array.go
func medianOfUniquenessArray(nums []int) int {
	n := len(nums)
	total := (n * (n + 1)) / 2
	left, right := 1, n
	answer := 1
	for left <= right {
		mid := (left + right) / 2
		count := checkCount(nums, mid)
		if count*2 >= total {
			answer = mid
			right = mid - 1
		} else {
			left = mid + 1
		}
	}
	return answer
}
func checkCount(nums []int, k int) int {
	answer := 0
	left := 0
	count := make(map[int]int)
	size := 0
	for right := range nums {
		count[nums[right]] += 1
		if count[nums[right]] == 1 {
			size += 1
		}
		for size > k {
			count[nums[left]] -= 1
			if count[nums[left]] == 0 {
				size -= 1
			}
			left += 1
		}
		answer += right - left + 1
	}
	return answer
}